lib/spack/spack/version.py

```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+            with open(self.repository_metadata_path, 'r') as fd:
+                self.data = sjson.load(fd.read())
+
+    def get_commit_lookup(self, version, spack_versions):
+        """
+        Given a git handle, repository, and known spack versions, generate a lookup.
+
+        We do this by listing all the tags for a repository path, in the order of
+        release (the assumption being that versioned releases/candidates that are
+        later are preferable to earlier) and then we match these versions to known
+        spack versions. We can then parse the entire git history, and generate
+        a lookup that describes the context of every commit to a known spack
+        version. In the case that commit is an exact match, exact_version will
+        be true. If a commit does not have a spack version before it, or does
+        not have a spack version after it, then the value will be None.
+        """
+        # lookup cache and return early here
+        self.load_data()
+
+        # Cut out early if we already have the version cached
+        if str(version) in self.data:
+            return self.data
+
+        # Were we given an already cloned path?
+        already_cloned = os.path.exists(self.repository_uri)
+        if already_cloned:
+            dest = self.repository_uri
+
+        # Otherwise honor the namespace for the service and repository
+        else:
+            dest = os.path.join(spack.paths.user_repos_cache_path, self.repository_uri)
+
+        # prepare a cache for the repository and metadata
+        for path in [dest, self.repository_metadata_path]:
+            parent = os.path.dirname(path)
+            if not os.path.exists(parent):
+                os.makedirs(parent)
+
+        # Only clone if we don't have it!
+        if not os.path.exists(dest):
+            self.fetcher.clone(os.path.dirname(dest))
+
+        # But if we get here and no repostiory, this is a fail
+        if not os.path.exists(dest):
+            tty.die("There was an issue cloning %s to %s" % (self.fetcher.url, dest))
+
+        # Always pull and fetch
+        with working_dir(dest):
+
+            # Don't do anything if already cloned - it might not have remote
+            if not already_cloned:
+                self.fetcher.git("pull")
+                self.fetcher.git("fetch")
+
+            # List tags (refs) by date, so last reference of a tag is newest
+            tags = self.fetcher.git("for-each-ref", "--sort=creatordate", "--format",
+                                    "%(refname)", "refs/tags", output=str).split('\n')
+
+        # Lookup of spack tags to git references
+        tag_lookup = {}
+
+        # Lookup of commits to spack versions
+        commit_to_version = {}
+
+        # For each tag, try to match to a version.
+        versions = [str(v) for v in spack_versions]
+        for tag in tags:
+            tag = tag.replace('refs/tags/', '', 1)
+            for v in versions:
+                if v in tag:
```

Comment:
... yes it would. And we don't have any good way to impose an ordering on the comparisons to help either.

I think for now I'll change to checking for the tag equaling the version or equaling 'v' + version.